### ISSUE TITLE
Passing structures to callbacks by-value

### ISSUE THREAD

**Issue by sionescu** [cffi/cffi#285]
Theres seems to be no way to pass structures to callbacks by value, even with libffi, although all other things support that with libffi.

Launchpad Details: [#LP1828643](https://bugs.launchpad.net/bugs/1828643) Ekaterina Vaartis - 2019-05-10 23:43:44 +0000

**Comment by sionescu** [comment]
This is indeed a big oversight. Is this something you'd be interested to work on?

Launchpad Details: [#LPC](https://bugs.launchpad.net/cffi/+bug/1828643/comments/1) Luís Oliveira - 2019-05-14 09:18:56 +0000

**Comment by sionescu** [comment]
If you can give me some guidance on how to actually do this, i could do it. I have no experience with libffi though.

Launchpad Details: [#LPC](https://bugs.launchpad.net/cffi/+bug/1828643/comments/2) Ekaterina Vaartis - 2019-05-14 10:37:56 +0000

**Comment by sionescu** [comment]
So, the first step is to learn how to set up callbacks using libffi. I'm guessing their Closure API is the way to go. You could do this in C first and then translate that to CFFI, or start with CFFI, depending on how familiar you are with CFFI. You definitely need to be familiar with C, I think. :-)

Launchpad Details: [#LPC](https://bugs.launchpad.net/cffi/+bug/1828643/comments/3) Luís Oliveira - 2019-05-14 12:29:00 +0000

**Comment by sionescu** [comment]
I'll try to check it out and will write back later.

Launchpad Details: [#LPC](https://bugs.launchpad.net/cffi/+bug/1828643/comments/4) Ekaterina Vaartis - 2019-05-14 12:58:16 +0000

**Comment by sionescu** [comment]
Although it'd be nice to also know how libffi integrates with CFFI, since the libffi system is pretty small and i don't see direct places that explain passing strcutures by-value.

Launchpad Details: [#LPC](https://bugs.launchpad.net/cffi/+bug/1828643/comments/5) Ekaterina Vaartis - 2019-05-14 13:05:40 +0000

**Comment by sionescu** [comment]
foreign-funcall-form/fsbv-with-libffi (
https://github.com/cffi/cffi/blob/master/libffi/funcall.lisp#L82) is the
main entry point.


Launchpad Details: [#LPC](https://bugs.launchpad.net/cffi/+bug/1828643/comments/6) Luís Oliveira - 2019-05-14 18:54:49 +0000

**Comment by sionescu** [comment]
It seems that `cffi:defcallback` doesn't currently use libffi for creation callbacks.
E.g. in case of sbcl, it calls `%defcallback`, that is defined in `cffi-sbcl.lisp` and calls SBCL's own `alien-sap`.

That means, if we want to add support for libffi callbacks, we need to define a new macro, like `cffi:defcallback-libffi` or something.

Launchpad Details: [#LPC](https://bugs.launchpad.net/cffi/+bug/1828643/comments/7) Alexey Egorov - 2019-06-01 08:17:36 +0000

**Comment by sionescu** [comment]
Alexey, that shouldn't be necessary. Have a look at how DEFCFUN (and FOREIGN-FUNCALL, etc) decides to use libffi when needed without resorting to an auxiliar DEFCFUN-LIBFFI.

Launchpad Details: [#LPC](https://bugs.launchpad.net/cffi/+bug/1828643/comments/8) Luís Oliveira - 2019-06-01 09:43:30 +0000

**Comment by Angel-Cortez** [comment]
@sionescu, I have branch that implements this functionality. Am I expected to implement a solution for every Common Lisp implementation for my patch to be merged? I currently only have a working SBCL implementation for this.

**Comment by luismbo** [comment]
Why is it SBCL-specific?

**Comment by Angel-Cortez** [comment]
> Why is it SBCL-specific?

It's just the implementation I'm most familiar with most compatible with the `cl-objc` at the moment. We should be able to architect a similar solution for the other implementations.

**Comment by Angel-Cortez** [comment]
@luismbo the pointer APIs are not standard. For example, creating a function pointer via `sb-alien::alien-lambda` in SBCL.

**Comment by luismbo** [comment]
@Angel-Cortez can you use `cffi:defcallback` instead?

**Comment by Angel-Cortez** [comment]
> @Angel-Cortez can you use `cffi:defcallback` instead?

@luismbo, good point my patch is still in draft. That should work. I'll report back if I find any issues.

**Comment by Angel-Cortez** [comment]
@luismbo, your suggestion did work. Thank you, I just need to do some clean up of the implementation and add some unit tests before I post an official PR.

**Comment by Angel-Cortez** [comment]
UPDATE: Still a work in progress. I need to do more clean up and add some more tests: #351
cc: @luismbo

**Comment by theangelperalta** [comment]
@luismbo, not sure if you have any additional feedback on this PR?. Is it good to merge?
